compiler/mir/mirpasses.nim

```diff
@@ -478,6 +487,38 @@ proc fixupCallArguments(tree: MirTree, config: ConfigRef,
       of AllNodeKinds - ArgumentNodes:
         unreachable()
 
+proc lowerSwap(tree: MirTree, changes: var Changeset) =
+  ## Lowers a ``swap(a, b)`` call into:
+  ##
+  ## ..code-block:: nim
+  ##
+  ##   let tmp = a
+  ##   a = b
+  ##   b = tmp
+  ##
+  ## where all assignments are shallow.
+  for i in search(tree, mSwap):
+    let
+      typ = tree[operand(tree, Operation i, 0)].typ
+      temp = MirNode(kind: mnkTemp, typ: typ, temp: changes.getTemp())
+    changes.seek(i)
+    changes.replaceMulti(buf):
+      buf.subTree MirNode(kind: mnkRegion):
+        # the temporary doesn't need to own the value, so use ``DefCursor``
+        buf.add opParamNode(0, typ)
+        buf.subTree MirNode(kind: mnkDefCursor):
+          buf.add temp
+        # we're just swapping the values, no full copy is needed
+        argBlock(buf):
+          chain(buf): opParam(0, typ) => tag(ekReassign) => name()
+          chain(buf): opParam(1, typ) => arg()
+        buf.add MirNode(kind: mnkFastAsgn)
+        argBlock(buf):
+          chain(buf): opParam(1, typ) => tag(ekReassign) => name()
+          chain(buf): emit(temp) => arg()
+        buf.add MirNode(kind: mnkFastAsgn)
+    changes.remove() # remove the 'void' sink
```

Comment:
Having to remove the `mnkVoid` node isn't pretty, and the same thing is going to come up for the lowering passes of other magics. I think it'd make sense to introduce the new `mnkInlined` (or similar) node which acts much like an `mnkRegion`, with the difference that an `mnkInlined` returns a value ('unit' when the replaced operator has a `void` return).